### PR TITLE
Avoid endless loop

### PR DESCRIPTION
**Pull request by imoldovan-intacct**
Of course, I did not think this through...
This pull request is related to https://github.com/apache/httpcomponents-core/pull/58 .

We might get into an endless loop for sizes which are in (Integer.MAX_VALUE - headRoom, Integer.MAX_VALUE] .

### CODE REVIEW

File: httpcore5/src/main/java/org/apache/hc/core5/http/impl/nio/ExpandableBuffer.java [hunk 1]
```diff
@@ -123,6 +123,10 @@ protected void expand() {
             // WARNING: This code assumes you are providing enough heap room with -Xmx.
             // source of inspiration: https://bugs.openjdk.java.net/browse/JDK-8059914
             newcapacity = Integer.MAX_VALUE - headRoom;
+            if (newcapacity == this.buffer.capacity()) {
```
Comment (ok2c) [inline httpcore5/src/main/java/org/apache/hc/core5/http/impl/nio/ExpandableBuffer.java]: @imoldovan-intacct  why not `newcapacity >= this.buffer.capacity()` ?

Comment (ok2c) [reply httpcore5/src/main/java/org/apache/hc/core5/http/impl/nio/ExpandableBuffer.java]: @imoldovan-intacct you might also want to rebase your branch off the upstream master.

Comment (imoldovan-intacct) [reply httpcore5/src/main/java/org/apache/hc/core5/http/impl/nio/ExpandableBuffer.java]: I did that.
I also did a beginner mistake :) . I'll try not to commit with too much coffee on board anymore.

Comment (imoldovan-intacct) [reply httpcore5/src/main/java/org/apache/hc/core5/http/impl/nio/ExpandableBuffer.java]: There is no way for this.buffer.capacity() to exceed Integer.MAX_VALUE - headRoom.
When we get to the point that they are equal => we already have touched maximum buffer capacity allowed once (MAX_VALUE - 8 in our case) and we are going inside an endless loop.

Comment (ok2c) [reply httpcore5/src/main/java/org/apache/hc/core5/http/impl/nio/ExpandableBuffer.java]: bq. There is no way for this.buffer.capacity() to exceed Integer.MAX_VALUE - headRoom
@imoldovan-intacct  Probably, but still looks a little safer

**Comment by ok2c** [comment]
@imoldovan-intacct I would rather use `BufferOverflowException` instead of `MessageConstraintException`

**Comment by imoldovan-intacct** [comment]
Thanks for the suggestions. It's all good now.
I do have some improvements in mind for the buffer allocation, but I preferred to stick to the burning issue for now.

**Pull request closed without merging**